fish_completions.go

```diff
@@ -137,49 +126,65 @@ function __%[1]s_prepare_completions
 
     __%[1]s_debug "nospace: $nospace, nofiles: $nofiles"
 
-    # Important not to quote the variable for count to work
-    set numComps (count $__%[1]s_comp_results)
-    __%[1]s_debug "numComps: $numComps"
-
-    if test $numComps -eq 1; and test $nospace -ne 0
-        # To support the "nospace" directive we trick the shell
-        # by outputting an extra, longer completion.
-        __%[1]s_debug "Adding second completion to perform nospace directive"
-        set --append __%[1]s_comp_results $__%[1]s_comp_results[1].
+    # If we want to prevent a space, or if file completion is NOT disabled,
+    # we need to count the number of valid completions.
+    # To do so, we will filter on prefix as the completions we have received
+    # may not already be filtered so as to allow fish to match on different
+    # criteria than the prefix.
+    if test $nospace -ne 0; or test $nofiles -eq 0
+        set prefix (commandline -t)
+        __%[1]s_debug "prefix: $prefix"
+
+        set completions
+        for comp in $__%[1]s_comp_results
+            if test (string match -e -r -- "^$prefix" "$comp")
+                set -a completions $comp
+            end
+        end
```

Comment:
In case this is relevant: the regex should be escaped, just in case, and this loop can be done in one line because `string match` accepts multiple arguments. I'd also eschew the `-e` option, I think that option was a mistake.

```fish
set prefix (commandline -t | string escape --style=regex)
...
set -l completions (string match -r -- "^$prefix.*" $__%[1]s_comp_results)
```

Reply:
Thanks for the improvements @krobelus !

> the regex should be escaped, just in case,

Sounds safer, I'll add this.

> and this loop can be done in one line because `string match` accepts multiple arguments.

Nice!  I admit I was upset at having to do the loop.  Now it can be removed!

> I'd also eschew the `-e` option, I think that option was a mistake.

Actually, according to the following doc, I understood that using `-e` was preferred to adding the suffix `.*`, so I went that route: https://fishshell.com/docs/current/cmds/string-match.html

Reply:
The `-e` thing is just an opinion, anything works of course. My point is, without it the regex is easier to understand by someone who doesn't know this flag.